Docstring tests

Madnex opened the pull request:
Fixed multiple mostly small pytest errors. The tests regarding the StanfordParser are set to be skipped:

- nltk/parse/stanford.py
- nltk/tag/stanford.py
- nltk/tokenize/stanford.py
- nltk/tokenize/stanford_segmenter.py

After this PR most of the files in the project should pass the pytest docstring tests.

jingbels committed:
fixed pytests

jingbels committed:
fixed more pytests

jingbels committed:
fixed more pytest and changed multiline pytest issues fixes for snowball.py and causal.py

jingbels committed:
fixed pytests (mainly multiline or rounding issues)

jingbels committed:
fixed treebank pytests, removed test for return_string=True (deprecated)

jingbels committed:
fixed destructive.py pytests, removed test for return_string=True (deprecated)

jingbels committed:
fixed pytest (rounding issues)

jingbels committed:
fixed pytest (initialised missing object)

jingbels committed:
fixed pytest (formatting issues)

jingbels committed:
fixed pytest (formatting issues)

jingbels committed:
fixed pytest (formatting issues)

jingbels committed:
Merge branch 'develop' of https://github.com/nltk/nltk into docstring-tests

jingbels committed:
added pytest +SKIP for deprecated module stanford

jingbels committed:
updated AUTHORS.md

tomaarsen commented on nltk/metrics/agreement.py:
```diff
@@ -50,10 +50,10 @@
-    0.8199999999999998...
+    >>> round(t.pi(), 5)
+    0.79953
+    >>> round(t.S(), 3)
+    0.82
```
This seems odd, rounding to 3 gives 2 values after the dot?

Madnex replied:
Yes, because if not rounded we have 0.8199999999999998 :P

tomaarsen replied:
Oooh, that's rough, then `0.82...` wouldn't work either (presumably). Rounding is acceptable then, otherwise the output (e.g. `0.819...`) may change in the future or on different devices (e.g. to `0.820...`).

ekaf replied:
Wouldn't _round(t.S(), 2)_ be more appropriate, then?

tomaarsen commented on nltk/corpus/__init__.py:
```diff
@@ -54,8 +54,8 @@
     >>> from nltk.corpus import brown
-    >>> print(", ".join(brown.words()))
-    The, Fulton, County, Grand, Jury, said, ...
+    >>> print(", ".join(brown.words()[:6])) # only first 6 words
+    The, Fulton, County, Grand, Jury, said
```
It was my understanding that `...` is valid for tests, as long as you use `# doctest: +ELLIPSIS`. That seems cleanest, especially for cases where the alternative is really long.

Madnex replied:
Okay, I agree that's the easier and cleaner solution. How do I update this? Do I have to create a new pull request?

tomaarsen replied:
No need! You can push new commits to the `docstring-tests` branch on your nltk fork, and they will automatically be included in this PR.

tomaarsen commented on nltk/corpus/reader/framenet.py:
```diff
@@ -1510,8 +1514,48 @@ def frame(self, fn_fid_or_fname, ignorekeys=[]):
         >>> f.ID
         256
         >>> # ensure non-ASCII character in definition doesn't trigger an encoding error:
-        >>> fn.frame('Imposing_obligation')
-        frame (1494): Imposing_obligation...
+        >>> fn.frame('Imposing_obligation') # doctest: +NORMALIZE_WHITESPACE
+        frame (1494): Imposing_obligation
+        <BLANKLINE>
+        [URL] https://framenet2.icsi.berkeley.edu/fnReports/data/frame/Imposing_obligation.xml
```
Again, can we use `# doctest: +ELLIPSIS` here?

tomaarsen commented on nltk/draw/util.py:
```diff
@@ -90,7 +90,7 @@ class CanvasWidget(metaclass=ABCMeta):
     ``__getitem__`` operator:
 
-        >>> cn['color']
+        >>> print(cn['color'])
```
Is this necessary? Does it not correctly convert to string otherwise or something?

Madnex replied:
When I remove the print() the test fails: 

`Failed example:
    cn['color']
Expected:
    red
Got:
    'red'
`

tomaarsen replied:
Oh, that makes sense! It's expecting a variable, not a string, because that's what we've put as the expected answer. Changing `red` to `"red"` should also work, but it does not matter much.

tomaarsen commented on nltk/probability.py:
```diff
@@ -2208,8 +2208,8 @@ class ConditionalProbDist(ConditionalProbDistI):
         >>> cpdist['passed'].max()
         'VBD'
-        >>> cpdist['passed'].prob('VBD')
-        0.423...
+        >>> round(cpdist['passed'].prob('VBD'), 3)
+        0.423
 
```
Again, can this be fixed with ellipses? It feels a bit cleaner to me.

tomaarsen commented on nltk/tag/brill_trainer.py:
```diff
@@ -123,8 +123,8 @@ def train(self, train_sents, max_rules=200, min_score=2, min_acc=None):
         >>> baseline = backoff #see NOTE1
-        >>> baseline.accuracy(gold_data) #doctest: +ELLIPSIS
-        0.2450142...
+        >>> round(baseline.accuracy(gold_data), 7)
+        0.2433862
```
I believe the prior only failed because it is actually e.g. `0.243...`

tomaarsen reviewed (commented):
I quickly added some comments. Most of them related to `...` and `# doctest: +ELLIPSIS`. I'm in favor of using `...` where possible rather than `round()`, slicing a string to e.g. only print 6 words, or putting the entire output in the doctest output.

@ekaf @stevenbird what are your thoughts on the two "styles"?

tomaarsen commented:
I resolved a quick merge conflict in `AUTHORS.md` in 91f20c0.

stevenbird commented:
@tomaarsen I'm with you in preferring ellipses in the output, rather than modifying the code in ways that no-one would do IRL

ekaf commented:
@tomaarsen (https://github.com/nltk/nltk/pull/3050#pullrequestreview-1104252744) I agree that your preferred style is cleaner and generally more robust.

jingbels committed:
changed docstring corrections by usage of ELLIPSIS and different roundings

jingbels committed:
changed docstring corrections by usage of ELLIPSIS and different roundings

jingbels committed:
fixed AUTHORS.md to be consistent

Madnex commented:
I updated the code according to the suggestions 👍

ekaf commented:
Indeed, this PR significantly reduces the number of doctest failures:

```

=========================== short test summary info ============================
FAILED nltk/text.py::nltk.text.Text.collocation_list
FAILED nltk/corpus/reader/framenet.py::nltk.corpus.reader.framenet.FramenetCorpusReader.lu
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.__init__
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.configure
FAILED nltk/draw/table.py::nltk.draw.table.Table
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPDependencyParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tag
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tokenize
FAILED nltk/parse/dependencygraph.py::nltk.parse.dependencygraph.DependencyGraph._repr_svg_
FAILED nltk/tag/hunpos.py::nltk.tag.hunpos.HunposTagger
FAILED nltk/test/unit/test_downloader.py::test_downloader_using_existing_parent_download_dir
FAILED nltk/test/unit/test_downloader.py::test_downloader_using_non_existing_parent_download_dir
FAILED nltk/tokenize/punkt.py::nltk.tokenize.punkt.PunktSentenceTokenizer._match_potential_end_contexts
= 14 failed, 681 passed, 33 skipped, 9 xfailed, 22 warnings in 414.32s (0:06:54) =

```

Note that #3048 fixes the remaining failures in _nltk/parse/dependencygraph.py_ and _nltk/tag/hunpos.py_.

ekaf commented:
There are 3 easy ways of solving the remaining failure in _nltk/corpus/reader/framenet.py_ :  either use _print_ instead of _pprint_ in line 1657, or doctest: +NORMALIZE_WHITESPACE, or format the expected test output with _pprint_:

```
________ [doctest] nltk.corpus.reader.framenet.FramenetCorpusReader.lu _________
1648         Usage examples:
1649 
1650         >>> from nltk.corpus import framenet as fn
1651         >>> fn.lu(256).name
1652         'foresee.v'
1653         >>> fn.lu(256).definition
1654         'COD: be aware of beforehand; predict.'
1655         >>> fn.lu(256).frame.name
1656         'Expectation'
1657         >>> pprint(list(map(PrettyDict, fn.lu(256).lexemes)))
Expected:
    [{'POS': 'V', 'breakBefore': 'false', 'headword': 'false', 'name': 'foresee', 'order': 1}]
Got:
    [{'POS': 'V',
      'breakBefore': 'false',
      'headword': 'false',
      'name': 'foresee',
      'order': 1}]

nltk/corpus/reader/framenet.py:1657: DocTestFailure

```

tomaarsen committed:
Merge branch 'develop' into docstring-tests

tomaarsen commented:
After merging @ekaf's #3048 and then checking out this PR, I get the following failing tests still:
```
====================================================================================== short test summary info ====================================================================================== 
FAILED nltk/text.py::nltk.text.Text.collocation_list
FAILED nltk/corpus/reader/framenet.py::nltk.corpus.reader.framenet.FramenetCorpusReader.lu
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.__init__
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.configure
FAILED nltk/draw/table.py::nltk.draw.table.Table
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPDependencyParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tag
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tokenize
FAILED nltk/parse/transitionparser.py::nltk.parse.transitionparser.demo
FAILED nltk/tokenize/punkt.py::nltk.tokenize.punkt.PunktSentenceTokenizer._match_potential_end_contexts
========================================================== 11 failed, 688 passed, 29 skipped, 9 xfailed, 22 warnings in 125.34s (0:02:05) ===========================================================
```
This differs slightly from the previous list. 

Do we want to get this merged now, and then look at those remaining tests in another issue/PR, or do we want to fix these in this PR too? I recognize that the remaining issues are a bit more challenging.

ekaf commented:
@tomaarsen (https://github.com/nltk/nltk/pull/3050#issuecomment-1251128936) the two additional failures with _nltk/test/unit/test_downloader.py_ in the previous list were due to being offline. When testing online I get the same list as you.

ekaf commented:
I don't think this PR should try to fix everything, but maybe the easy framenet.py fix could fit here.

tomaarsen committed:
Fix framenet doctest formatting with pprint

tomaarsen commented:
As suggested by @ekaf, I've quickly fixed the framenet test. The failing tests are now:
```
====================================================================================== short test summary info ====================================================================================== 
FAILED nltk/text.py::nltk.text.Text.collocation_list
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.__init__
FAILED nltk/draw/table.py::nltk.draw.table.MultiListbox.configure
FAILED nltk/draw/table.py::nltk.draw.table.Table
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPDependencyParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.CoreNLPParser
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tag
FAILED nltk/parse/corenlp.py::nltk.parse.corenlp.GenericCoreNLPParser.tokenize
FAILED nltk/parse/transitionparser.py::nltk.parse.transitionparser.demo
FAILED nltk/tokenize/punkt.py::nltk.tokenize.punkt.PunktSentenceTokenizer._match_potential_end_contexts
========================================================== 10 failed, 689 passed, 29 skipped, 9 xfailed, 22 warnings in 143.19s (0:02:23) =========================================================== 
```
I'll merge this once the tests go green. Thanks for the work @Madnex & @ekaf!

tomaarsen committed:
Change docstring on MultiListBox.__init__

I believe the original typo was misinterpreted and changed to something that was not originally intended.
